src/gwt/src/org/rstudio/core/client/command/EditorCommandManager.java

```diff
@@ -0,0 +1,176 @@
+      
+      private final native String getBindingString()
+      /*-{
+         return this;
+      }-*/;
+   }
+   
+   public EditorCommandManager()
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      
+      bindings_ = new FileBacked<EditorKeyBindings>(
+            KEYBINDINGS_PATH,
+            EditorKeyBindings.create());
+      
+      events_.addHandler(
+            EditorLoadedEvent.TYPE,
+            new EditorLoadedHandler()
+            {
+               @Override
+               public void onEditorLoaded(EditorLoadedEvent event)
+               {
+                  loadBindings();
+               }
+            });
+   }
+   
+   @Inject
+   private void initialize(EventBus events,
+                           FilesServerOperations files)
+   {
+      events_ = events;
+      files_ = files;
+   }
+   
+   public static final native JsArray<AceCommand> getDefaultAceCommands() /*-{
+      return $wnd.require("ace/commands/default_commands").commands;
+   }-*/;
+   
+   private AceCommandManager getAceCommandManager()
+   {
+      RequestEditorCommandsEvent event = new RequestEditorCommandsEvent();
+      events_.fireEvent(event);
+      return event.getAceCommandManager();
+   }
+   
+   public boolean hasBinding(KeySequence keys)
+   {
+      AceCommandManager manager = getAceCommandManager();
+      return manager.hasBinding(keys);
+   }
+   
+   public boolean hasPrefix(KeySequence keys)
+   {
+      AceCommandManager manager = getAceCommandManager();
+      return manager.hasPrefix(keys);
+   }
+   
+   public void rebindCommand(String id, KeySequence keySequence)
+   {
+      events_.fireEvent(new AddEditorCommandEvent(id, keySequence, true));
+   }
+   
+   public void addBindingsAndSave(final EditorKeyBindings newBindings)
+   {
+      bindings_.execute(new CommandWithArg<EditorKeyBindings>()
+      {
+         @Override
+         public void execute(final EditorKeyBindings currentBindings)
+         {
+            currentBindings.insert(newBindings);
+            bindings_.set(currentBindings, new Command()
+            {
+               @Override
+               public void execute()
+               {
+                  loadBindings();
+               }
+            });
+         }
+      });
+   }
+   
+   public void loadBindings()
+   {
+      bindings_.execute(new CommandWithArg<EditorKeyBindings>()
+      {
+         @Override
+         public void execute(EditorKeyBindings bindings)
+         {
+            for (String commandName : bindings.iterableKeys())
+            {
+               EditorKeyBinding binding = bindings.get(commandName);
+               rebindCommand(
+                     commandName,
+                     binding.getKeyBinding());
+            }
+         }
+      });
+   }
+   
+   private final FileBacked<EditorKeyBindings> bindings_;
+   private boolean isBindingsLoaded_ = false;
+   public static final String KEYBINDINGS_PATH = "~/.R/keybindings/editor_bindings.json";
```

Comment:
Yes, I think for IDE specific things we should use e.g. `~/.R/rstudio/keybindings` (snippets aren't in that category to me because other editors could support them)